### PR TITLE
Revamp icon legend as a modal

### PR DESCRIPTION
**Pull request by janfaracik**
image

* Introduces a new lightweight modal component (currently lacking in depth, but eventually it could be used for confirmation dialogs, notices etc)
* Replaces the icon legend page with an accessible modal

Relates to https://issues.jenkins.io/browse/JENKINS-68357

### Testing done

* Icon legend works the same as before

### Proposed changelog entries

- Revamp icon legend as a modal

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7718"></a>

### CODE REVIEW

**Comment by timja** [comment]
Looks really nice.

Feedback / questions
1. Should an overlay be placed on the rest of the page to show you can't interact with it when the modal is open?
2. Should clicking outside the modal close it? (It currently doesn't)
3. Contrast isn't sufficient for disabled / aborted on dark theme:

image

**Comment by janfaracik** [comment]
> 1. Should an overlay be placed on the rest of the page to show you can't interact with it when the modal is open?

There is an overlay (it darkens the page)

> 2. Should clicking outside the modal close it? (It currently doesn't)

That makes sense to me, especially for a simple modal like this.

> 3. Contrast isn't sufficient for disabled / aborted on dark theme:

👍 Just opened a PR for dark theme https://github.com/jenkinsci/dark-theme-plugin/pull/338

Thanks!

**Comment by timja** [comment]
> There is an overlay (it darkens the page)

The overlay only works on safari, doesn't work on Firefox or Chrome

**Comment by janfaracik** [comment]
> > There is an overlay (it darkens the page)
> 
> The overlay only works on safari, doesn't work on Firefox or Chrome

Hadn't spotted that, thanks - opened a new dark theme PR to address https://github.com/janfaracik/dark-theme-plugin/pull/44

**Comment by timja** [comment]
That’s been opened against your fork not the origin

**Comment by janfaracik** [comment]
> That’s been opened against your fork not the origin

Whoops, opened it again https://github.com/jenkinsci/dark-theme-plugin/pull/339

File: war/src/main/webapp/scripts/prototype.js
Comment (timja) [inline war/src/main/webapp/scripts/prototype.js:1993]: why was this removed? it's what is causing the ATH failures.

Comment (janfaracik) [reply war/src/main/webapp/scripts/prototype.js:1993]: It was interfering with the removal of the modal (I'm assuming because the modal doesn't have a parent?

Comment (timja) [reply war/src/main/webapp/scripts/prototype.js:1993]: do we still need to remove this function now that you've changed the core components to not use it?
https://github.com/search?q=org%3Ajenkinsci+Element.remove%28+language%3AJavaScript&type=code&l=JavaScript&p=1

It looks like it _should_ be okay to remove.

Comment (timja) [reply war/src/main/webapp/scripts/prototype.js:1993]: (Accidentally resolved this in case you missed it)

Comment (janfaracik) [reply war/src/main/webapp/scripts/prototype.js:1993]: If I'm understanding correctly, we still need to remove the function as it interferes with removing the modal.

Comment (timja) [reply war/src/main/webapp/scripts/prototype.js:1993]: I tested locally and it seems to be working fine? not sure how it could be causing an issue if nothing is calling it.

Comment (janfaracik) [reply war/src/main/webapp/scripts/prototype.js:1993]: image

With the remove() function restored I get the above error when closing the modal.

Comment (timja) [reply war/src/main/webapp/scripts/prototype.js:1993]: weird I get that error in Safari but not chrome...

Maybe it has a parent in chrome, I guess we already hit a browser difference with this functionality.

**Review by NotMyFault — APPROVED** [review]
A very welcoming change :)

**Review by yaroslavafenkin — APPROVED** [review]
Tested manually on FF, Chrome and Safari, seems to behave as expected.
Security-wise looks good to me too.

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/resources/jenkins/model/Jenkins/legend.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/legend.jelly:28]: Since this is no longer a standalone view, attempts to access the URL `/legend` (e.g. from links in outside documentation, or browser history) will result in the logging of the message

```
Mar 17, 2023 5:00:06 PM WARNING jenkins.security.stapler.StaplerFilteredActionListener onDispatchTrigger
New Stapler dispatch rules result in the URL "/legend" no longer being allowed. If you consider it safe to use, add the following to the whitelist: "hudson.model.Hudson legend". Learn more: https://www.jenkins.io/redirect/stapler-facet-restrictions
```

This seems unnecessary. Perhaps rename the view to `_legend` so it's a simple 404?

Comment (timja) [reply core/src/main/resources/jenkins/model/Jenkins/legend.jelly:28]: https://github.com/jenkinsci/jenkins/pull/7746

**Pull request merged**
